src/jit/flowgraph.cpp

```diff
@@ -8159,71 +8159,40 @@ void                Compiler::fgAddInternal()
-        (varTypeIsStruct(info.compRetNativeType) && info.compRetBuffArg == BAD_VAR_NUM)))
-#else // !defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
-    if (genReturnBB && (info.compRetType != TYP_VOID) && !varTypeIsStruct(info.compRetNativeType))
-#endif // !defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
+    // it'll be reloaded by the single return.                                    
+    if (genReturnBB && compMethodHasRetVal())
     {
         genReturnLocal = lvaGrabTemp(true DEBUGARG("Single return block return value"));
-#if defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
-        var_types retLocalType = TYP_STRUCT;
-        if (info.compRetNativeType == TYP_STRUCT)
+
+        if (compMethodReturnsNativeScalarType())
         {
-            // If the native ret type is a struct, make sure the right 
-            // normalized type is assigned to the local variable.
-            SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR structDesc;
-            assert(info.compMethodInfo->args.retTypeClass != nullptr);
-            eeGetSystemVAmd64PassStructInRegisterDescriptor(info.compMethodInfo->args.retTypeClass, &structDesc);
-            if (structDesc.passedInRegisters && structDesc.eightByteCount <= 1)
-            {
-                retLocalType = lvaTable[genReturnLocal].lvType = getEightByteType(structDesc, 0);
-            }
-            else
-            {
-                lvaTable[genReturnLocal].lvType = TYP_STRUCT;
-            }
+            lvaTable[genReturnLocal].lvType = genActualType(info.compRetNativeType);
         }
-        else
-#endif // defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
+        else if (compMethodReturnsRetBufAddr())
         {
-            lvaTable[genReturnLocal].lvType = genActualType(info.compRetNativeType);
+            lvaTable[genReturnLocal].lvType = TYP_BYREF;
         }
-        
-        if (varTypeIsFloating(lvaTable[genReturnLocal].lvType))
+        else if (compMethodReturnsMultiRegRetType())
         {
-            this->compFloatingPointUsed = true;
+            lvaTable[genReturnLocal].lvType = TYP_STRUCT;
+            lvaSetStruct(genReturnLocal, info.compMethodInfo->args.retTypeClass, true);
+#if FEATURE_MULTIREG_ARGS_OR_RET
+            lvaTable[genReturnLocal].lvIsMultiRegArgOrRet = true;
+#endif
         }
-
-#if defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
-        // Handle a struct return type for System V Amd64 systems.
-        if (varTypeIsStruct(info.compRetNativeType))
+        else
         {
-            // Handle the normalized return type.
-            if (varTypeIsStruct(retLocalType))
-            {
-                lvaSetStruct(genReturnLocal, info.compMethodInfo->args.retTypeClass, true);
-            }
-            else
-            {
-                lvaTable[genReturnLocal].lvVerTypeInfo = typeInfo(TI_STRUCT, info.compMethodInfo->args.retTypeClass);
-            }
+            assert(!"unreached");
```

Comment:
Minor suggestion - but IMO it would be cleaner to remove the condition from the last "else if", and then add `assert(compMethodReturnsMultiRegRetType());`